pilot/pkg/xds/ecds.go

```diff
@@ -59,7 +65,44 @@ func (e *EcdsGenerator) Generate(proxy *model.Proxy, push *model.PushContext, w
 	if !ecdsNeedsPush(req) {
 		return nil, model.DefaultXdsLogDetails, nil
 	}
-	ec := e.Server.ConfigGenerator.BuildExtensionConfiguration(proxy, push, w.ResourceNames)
+
+	secretResources := referencedSecrets(proxy, push, w.ResourceNames)
+	var updatedSecrets map[model.ConfigKey]struct{}
+	// Check if the secret updates is relevant to Wasm image pull. If not relevant, skip pushing ECDS.
+	if !model.ConfigsHaveKind(req.ConfigsUpdated, gvk.WasmPlugin) && !model.ConfigsHaveKind(req.ConfigsUpdated, gvk.EnvoyFilter) &&
+		model.ConfigsHaveKind(req.ConfigsUpdated, gvk.Secret) {
+		// Get the updated secrets
+		updatedSecrets = model.ConfigsOfKind(req.ConfigsUpdated, gvk.Secret)
+		needsPush := false
+		for _, sr := range secretResources {
+			if _, found := updatedSecrets[model.ConfigKey{Kind: gvk.Secret, Name: sr.Name, Namespace: sr.Namespace}]; found {
+				needsPush = true
+				break
+			}
+		}
+		if !needsPush {
+			return nil, model.DefaultXdsLogDetails, nil
+		}
+	}
+
+	// Generate the pull secrets first, which will be used when populating the extension config.
+	var secretController credscontroller.Controller
+	if e.secretController != nil {
+		var err error
+		secretController, err = e.secretController.ForCluster(proxy.Metadata.ClusterID)
+		if err != nil {
+			log.Warnf("proxy %s is from an unknown cluster, cannot retrieve certificates for Wasm image pull: %v", proxy.ID, err)
+			return nil, model.DefaultXdsLogDetails, nil
+		}
+	}
+	var secrets map[string][]byte
+	if len(secretResources) > 0 {
+		// Inserts Wasm pull secrets in ECDS response, which will be used at xds proxy for image pull.
+		// Before forwarding to Envoy, xds proxy will remove the secret from ECDS response.
+		secrets = e.GeneratePullSecrets(proxy, updatedSecrets, secretResources, secretController, req)
```

Comment:
Currently it seems we fail open if we cannot fetch secret. IE we send the extension without a secret. It seems like maybe we should fail closed..? wdyt?

Reply:
I think it is fine to fail open, which should result in a fetch failure at proxy, which would be more visible to the user.